File: baseLib/src/androidTest/java/me/ycdev/android/lib/common/provider/InfoProviderClientTest.java
package me.ycdev.android.lib.common.provider;

import android.database.ContentObserver;
import android.os.Handler;
import android.os.Looper;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import static com.google.common.truth.Truth.assertThat;

@RunWith(AndroidJUnit4.class)
public class InfoProviderClientTest {
    private static final String TABLE_TEST = "test";
    private static final String KEY_STR_1 = "key_str1";
    private static final String KEY_BOOL_1 = "key_bool1";
    private static final String KEY_INT_1 = "key_int1";

    private InfoProviderClient mInfoClient;

    @Before
    public void setup() {
        mInfoClient = new InfoProviderClient(ApplicationProvider.getApplicationContext(),
                "me.ycdev.android.lib.common.provider.InfoProvider");

        // clear dirty data
        mInfoClient.remove(TABLE_TEST, KEY_STR_1);
        mInfoClient.remove(TABLE_TEST, KEY_BOOL_1);
        mInfoClient.remove(TABLE_TEST, KEY_INT_1);
    }

    @Test
    public void getAndPutString() {
        String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, "");
        assertThat(value).isEqualTo("");

        value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isNull();

        boolean result = mInfoClient.putString(TABLE_TEST, KEY_STR_1, "value_str1");
        assertThat(result).isTrue();

        value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isEqualTo("value_str1");
    }

    @Test
    public void getAndPutBoolean() {
        boolean value = mInfoClient.getBoolean(TABLE_TEST, KEY_BOOL_1, true);
        assertThat(value).isTrue();

        value = mInfoClient.getBoolean(TABLE_TEST, KEY_BOOL_1, false);
        assertThat(value).isFalse();

        boolean result = mInfoClient.putBoolean(TABLE_TEST, KEY_BOOL_1, true);
        assertThat(result).isTrue();

        value = mInfoClient.getBoolean(TABLE_TEST, KEY_BOOL_1, false);
        assertThat(value).isTrue();
    }

    @Test
    public void getAndPutInt() {
        int value = mInfoClient.getInt(TABLE_TEST, KEY_INT_1, 1);
        assertThat(value).isEqualTo(1);

        value = mInfoClient.getInt(TABLE_TEST, KEY_INT_1, 2);
        assertThat(value).isEqualTo(2);

        boolean result = mInfoClient.putInt(TABLE_TEST, KEY_INT_1, 3);
        assertThat(result).isTrue();

        value = mInfoClient.getInt(TABLE_TEST, KEY_INT_1, 1);
        assertThat(value).isEqualTo(3);
    }

    @Test
    public void putAndAddContentObserver() throws InterruptedException {
        String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isNull();

        final CountDownLatch latch = new CountDownLatch(1);
        ContentObserver observer = new ContentObserver(new Handler(Looper.getMainLooper())) {
            @Override
            public void onChange(boolean selfChange) {
                String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
                assertThat(value).isEqualTo("value_str1");
                latch.countDown();
            }
        };
        mInfoClient.registerObserver(TABLE_TEST, KEY_STR_1, observer);

        boolean result = mInfoClient.putString(TABLE_TEST, KEY_STR_1, "value_str1");
        assertThat(result).isTrue();

        value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isEqualTo("value_str1");

        latch.await();
        mInfoClient.unregisterObserver(observer);
    }

    @Test
    public void removeAndAddContentObserver() throws InterruptedException {
        boolean result = mInfoClient.putString(TABLE_TEST, KEY_STR_1, "value_str1");
        assertThat(result).isTrue();

        String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isEqualTo("value_str1");

        final CountDownLatch latch = new CountDownLatch(1);
        ContentObserver observer = new ContentObserver(new Handler(Looper.getMainLooper())) {
            @Override
            public void onChange(boolean selfChange) {
                String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
                assertThat(value).isNull();
                latch.countDown();
            }
        };
        mInfoClient.registerObserver(TABLE_TEST, KEY_STR_1, observer);

        result = mInfoClient.remove(TABLE_TEST, KEY_STR_1);
        assertThat(result).isTrue();

        value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isNull();

        latch.await();
        mInfoClient.unregisterObserver(observer);
    }

    @Test
    public void putButNoChange() throws InterruptedException {
        String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isNull();

        final CountDownLatch latch = new CountDownLatch(2);
        ContentObserver observer = new ContentObserver(new Handler(Looper.getMainLooper())) {
            @Override
            public void onChange(boolean selfChange) {
                String value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
                assertThat(value).isEqualTo("value_str1");
                latch.countDown();
            }
        };
        mInfoClient.registerObserver(TABLE_TEST, KEY_STR_1, observer);

        // put 1
        boolean result = mInfoClient.putString(TABLE_TEST, KEY_STR_1, "value_str1");
        assertThat(result).isTrue();

        value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isEqualTo("value_str1");

        // put 2 (but no value change)
        result = mInfoClient.putString(TABLE_TEST, KEY_STR_1, "value_str1");
        assertThat(result).isTrue();

        value = mInfoClient.getString(TABLE_TEST, KEY_STR_1, null);
        assertThat(value).isEqualTo("value_str1");

        latch.await(5, TimeUnit.SECONDS);
        assertThat(latch.getCount()).isEqualTo(1);

        mInfoClient.unregisterObserver(observer);
    }
}

File: baseLib/src/androidTest/java/me/ycdev/android/lib/common/provider/InfoProviderClientTest.kt
package me.ycdev.android.lib.common.provider

import android.database.ContentObserver
import android.os.Handler
import android.os.Looper
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.common.truth.Truth.assertThat
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class InfoProviderClientTest {
    private var mInfoClient: InfoProviderClient? = null

    @Before
    fun setup() {
        mInfoClient =
            InfoProviderClient(
                ApplicationProvider.getApplicationContext(),
                "me.ycdev.android.lib.common.provider.InfoProvider"
            )

        // clear dirty data
        mInfoClient!!.remove(TABLE_TEST, KEY_STR_1)
        mInfoClient!!.remove(TABLE_TEST, KEY_BOOL_1)
        mInfoClient!!.remove(TABLE_TEST, KEY_INT_1)
        mInfoClient!!.remove(null, KEY_DEFAULT_STR)
        mInfoClient!!.remove(TABLE_OTHER, KEY_STR_1)
        mInfoClient!!.remove(TABLE_TEST, KEY_INVALID_INT)
        mInfoClient!!.remove(TABLE_TEST, KEY_MISSING)
    }

    @Test
    fun getAndPutString() {
        var value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, "")
        assertThat(value).isEqualTo("")

        value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isNull()

        val result = mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value_str1")
        assertThat(result).isTrue()

        value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isEqualTo("value_str1")
    }

    @Test
    fun getAndPutBoolean() {
        var value = mInfoClient!!.getBoolean(TABLE_TEST, KEY_BOOL_1, true)
        assertThat(value).isTrue()

        value = mInfoClient!!.getBoolean(TABLE_TEST, KEY_BOOL_1, false)
        assertThat(value).isFalse()

        val result = mInfoClient!!.putBoolean(TABLE_TEST, KEY_BOOL_1, true)
        assertThat(result).isTrue()

        value = mInfoClient!!.getBoolean(TABLE_TEST, KEY_BOOL_1, false)
        assertThat(value).isTrue()
    }

    @Test
    fun getAndPutInt() {
        var value = mInfoClient!!.getInt(TABLE_TEST, KEY_INT_1, 1)
        assertThat(value).isEqualTo(1)

        value = mInfoClient!!.getInt(TABLE_TEST, KEY_INT_1, 2)
        assertThat(value).isEqualTo(2)

        val result = mInfoClient!!.putInt(TABLE_TEST, KEY_INT_1, 3)
        assertThat(result).isTrue()

        value = mInfoClient!!.getInt(TABLE_TEST, KEY_INT_1, 1)
        assertThat(value).isEqualTo(3)
    }

    @Test
    fun getAndPutString_defaultTable() {
        var value = mInfoClient!!.getString(null, KEY_DEFAULT_STR, "fallback")
        assertThat(value).isEqualTo("fallback")

        val result = mInfoClient!!.putString(null, KEY_DEFAULT_STR, "default-value")
        assertThat(result).isTrue()

        value = mInfoClient!!.getString(null, KEY_DEFAULT_STR, null)
        assertThat(value).isEqualTo("default-value")
    }

    @Test
    fun getAndPutString_isolatedAcrossTables() {
        assertThat(mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value-test")).isTrue()
        assertThat(mInfoClient!!.putString(TABLE_OTHER, KEY_STR_1, "value-other")).isTrue()

        assertThat(mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)).isEqualTo("value-test")
        assertThat(mInfoClient!!.getString(TABLE_OTHER, KEY_STR_1, null)).isEqualTo("value-other")
    }

    @Test
    fun getInt_invalidValueReturnsDefault() {
        assertThat(mInfoClient!!.putString(TABLE_TEST, KEY_INVALID_INT, "not-an-int")).isTrue()

        val value = mInfoClient!!.getInt(TABLE_TEST, KEY_INVALID_INT, 42)

        assertThat(value).isEqualTo(42)
    }

    @Test
    @Throws(InterruptedException::class)
    fun putAndAddContentObserver() {
        var value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isNull()

        val latch = CountDownLatch(1)
        val observer =
            object : ContentObserver(Handler(Looper.getMainLooper())) {
                override fun onChange(selfChange: Boolean) {
                    val curValue = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
                    assertThat(curValue).isEqualTo("value_str1")
                    latch.countDown()
                }
            }
        mInfoClient!!.registerObserver(TABLE_TEST, KEY_STR_1, observer)

        val result = mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value_str1")
        assertThat(result).isTrue()

        value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isEqualTo("value_str1")

        assertThat(latch.await(5, TimeUnit.SECONDS)).isTrue()
        mInfoClient!!.unregisterObserver(observer)
    }

    @Test
    @Throws(InterruptedException::class)
    fun removeAndAddContentObserver() {
        var result = mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value_str1")
        assertThat(result).isTrue()

        var value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isEqualTo("value_str1")

        val latch = CountDownLatch(1)
        val observer =
            object : ContentObserver(Handler(Looper.getMainLooper())) {
                override fun onChange(selfChange: Boolean) {
                    val curValue = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
                    assertThat(curValue).isNull()
                    latch.countDown()
                }
            }
        mInfoClient!!.registerObserver(TABLE_TEST, KEY_STR_1, observer)

        result = mInfoClient!!.remove(TABLE_TEST, KEY_STR_1)
        assertThat(result).isTrue()

        value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isNull()

        assertThat(latch.await(5, TimeUnit.SECONDS)).isTrue()
        mInfoClient!!.unregisterObserver(observer)
    }

    @Test
    @Throws(InterruptedException::class)
    fun putButNoChange() {
        var value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isNull()

        val latch = CountDownLatch(2)
        val observer =
            object : ContentObserver(Handler(Looper.getMainLooper())) {
                override fun onChange(selfChange: Boolean) {
                    val curValue = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
                    assertThat(curValue).isEqualTo("value_str1")
                    latch.countDown()
                }
            }
        mInfoClient!!.registerObserver(TABLE_TEST, KEY_STR_1, observer)

        // put 1
        var result = mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value_str1")
        assertThat(result).isTrue()

        value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isEqualTo("value_str1")

        // put 2 (but no value change)
        result = mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value_str1")
        assertThat(result).isTrue()

        value = mInfoClient!!.getString(TABLE_TEST, KEY_STR_1, null)
        assertThat(value).isEqualTo("value_str1")

        assertThat(latch.await(5, TimeUnit.SECONDS)).isFalse()
        assertThat(latch.count).isEqualTo(1)

        mInfoClient!!.unregisterObserver(observer)
    }

    @Test
    @Throws(InterruptedException::class)
    fun removeMissingValueDoesNotNotifyObserver() {
        val latch = CountDownLatch(1)
        val observer =
            object : ContentObserver(Handler(Looper.getMainLooper())) {
                override fun onChange(selfChange: Boolean) {
                    latch.countDown()
                }
            }
        mInfoClient!!.registerObserver(TABLE_TEST, KEY_MISSING, observer)

        val result = mInfoClient!!.remove(TABLE_TEST, KEY_MISSING)

        assertThat(result).isTrue()
        assertThat(latch.await(NO_CHANGE_TIMEOUT_MS, TimeUnit.MILLISECONDS)).isFalse()
        mInfoClient!!.unregisterObserver(observer)
    }

    @Test
    @Throws(InterruptedException::class)
    fun unregisterObserverStopsNotifications() {
        val latch = CountDownLatch(1)
        val observer =
            object : ContentObserver(Handler(Looper.getMainLooper())) {
                override fun onChange(selfChange: Boolean) {
                    latch.countDown()
                }
            }
        mInfoClient!!.registerObserver(TABLE_TEST, KEY_STR_1, observer)
        mInfoClient!!.unregisterObserver(observer)

        val result = mInfoClient!!.putString(TABLE_TEST, KEY_STR_1, "value_str1")

        assertThat(result).isTrue()
        assertThat(latch.await(NO_CHANGE_TIMEOUT_MS, TimeUnit.MILLISECONDS)).isFalse()
    }

    companion object {
        private const val NO_CHANGE_TIMEOUT_MS = 300L
        private const val TABLE_TEST = "test"
        private const val TABLE_OTHER = "other"
        private const val KEY_STR_1 = "key_str1"
        private const val KEY_DEFAULT_STR = "key_default_str"
        private const val KEY_BOOL_1 = "key_bool1"
        private const val KEY_INT_1 = "key_int1"
        private const val KEY_INVALID_INT = "key_invalid_int"
        private const val KEY_MISSING = "key_missing"
    }
}
